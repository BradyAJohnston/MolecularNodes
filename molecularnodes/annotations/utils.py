import inspect
import types
import typing
import bpy
from PIL import Image


def get_all_class_annotations(cls) -> dict:
    """Get all the annotations from the class including base classes"""
    all_annotations = {}
    for base_cls in cls.mro():
        if hasattr(base_cls, "__annotations__"):
            current_annotations = inspect.get_annotations(base_cls, eval_str=True)
            all_annotations.update(current_annotations)
    if "interface" in all_annotations:
        del all_annotations["interface"]
    return all_annotations


def get_blender_supported_type(atype):
    supported_types = (
        str,
        bool,
        int,
        float,
        tuple[float, float],
        tuple[float, float, float],
        tuple[float, float, float, float],
    )
    if atype in supported_types:
        return atype
    if typing.get_origin(atype) is typing.Union or type(atype) is types.UnionType:
        for utype in atype.__args__:
            if utype in supported_types:
                return utype
    return None


def get_view_matrix(obj):
    if obj._render_mode:
        return obj._scene.camera.matrix_world.inverted()
    else:
        return obj._rv3d.view_matrix


def is_perspective_projection(obj):
    if obj._render_mode:
        return obj._scene.camera.data.type != "ORTHO"
    else:
        return obj._rv3d.is_perspective


def render_annotations(
    scene: bpy.types.Scene, render_scale: float, image: Image, image_scale: float
) -> None:
    """Render annotations of all entities to an image"""
    session = scene.MNSession
    session.prune()  # remove any invalid session entities
    for entity in session.entities.values():
        if not hasattr(entity, "annotations"):
            continue
        manager = entity.annotations
        manager._enable_render_mode(scene, render_scale, image, image_scale)
        manager._draw_annotations()
        manager._disable_render_mode()
