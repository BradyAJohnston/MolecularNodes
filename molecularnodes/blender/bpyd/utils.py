import numpy as np


def centre(position: np.ndarray, weight: np.ndarray | None = None):
    "Calculate the weighted centroid of the vectors"
    if weight is None:
        return np.mean(position, axis=0)
    return np.sum(position * weight.reshape((-1, 1)), axis=0) / np.sum(weight)
