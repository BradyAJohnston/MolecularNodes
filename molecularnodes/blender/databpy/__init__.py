from .object import ObjectTracker, BlenderObject, bob, create_object
from .attribute import (
    named_attribute,
    store_named_attribute,
    Attribute,
    AttributeType,
    AttributeTypeInfo,
    AttributeTypes,
    Domains,
    DomainType,
)
