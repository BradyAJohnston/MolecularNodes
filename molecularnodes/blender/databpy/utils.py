import bpy


def evaluate_object(obj: bpy.types.Object):
    "Return an object which has the modifiers evaluated."
    obj.update_tag()
    return obj.evaluated_get(bpy.context.evaluated_depsgraph_get())
