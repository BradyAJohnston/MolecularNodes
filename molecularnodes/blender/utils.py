import bpy
from pathlib import Path


def path_resolve(path: str | Path) -> Path:
    if isinstance(path, str):
        return Path(bpy.path.abspath(path))
    elif isinstance(path, Path):
        return Path(bpy.path.abspath(str(path)))
    else:
        raise ValueError(f"Unable to resolve path: {path}")
