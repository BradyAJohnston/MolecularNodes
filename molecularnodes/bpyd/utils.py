import numpy as np


def centre(position: np.ndarray, weight: np.ndarray | None = None):
    "Calculate the weighted centroid of the vectors"
    if weight is None:
        return np.mean(position, axis=0)
    return np.sum(position * weight.reshape((-1, 1)), axis=0) / np.sum(weight)


def lerp(a: np.ndarray, b: np.ndarray, t: float = 0.5) -> np.ndarray:
    """
    Linearly interpolate between two values.

    Parameters
    ----------
    a : array_like
        The starting value.
    b : array_like
        The ending value.
    t : float, optional
        The interpolation parameter. Default is 0.5.

    Returns
    -------
    array_like
        The interpolated value(s).

    Notes
    -----
    This function performs linear interpolation between `a` and `b` using the
    interpolation parameter `t` such that the result lies between `a` and `b`.

    Examples
    --------
    >>> lerp(1, 2, 0.5)
    1.5

    >>> lerp(3, 7, 0.2)
    3.8

    >>> lerp([1, 2, 3], [4, 5, 6], 0.5)
    array([2.5, 3.5, 4.5])

    """
    return np.add(a, np.multiply(np.subtract(b, a), t))
