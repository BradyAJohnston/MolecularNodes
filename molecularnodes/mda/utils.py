"""
Utils
"""

import bpy
from bpy.app.handlers import persistent  # type: ignore


def is_mda_universe(object: bpy.types.Object) -> bool:
    """Check if blender object is MDA universe"""
    return object and object.mn.mda and object.mn.mda.is_mda_universe


def _msgbus_active_object_callback(*args):
    """Callback when active object changes"""
    sprop = bpy.context.scene.mn.mda
    active_object = bpy.context.active_object
    index = -1
    if is_mda_universe(active_object):
        # find index of universe in scene universes collection
        index = sprop.universes.find(active_object.mn.mda.universe_key)
    sprop.active_index = index  # updates selection in Universes panel


def _subscribe_to_active_object():
    """Actual subscribe to active object changes"""
    subscribe_to = (bpy.types.LayerObjects, "active")
    bpy.msgbus.subscribe_rna(
        key=subscribe_to,
        owner=bpy,
        args=(),
        notify=_msgbus_active_object_callback,
    )


@persistent
def active_object_load_handler(dummy):
    """Load handler to re-subscribe to active object changes"""
    _subscribe_to_active_object()


def subscribe_to_active_object():
    """Subscribe to active object changes"""
    _subscribe_to_active_object()  # subscribe
    # create a load handler to re-subscribe on file loads
    bpy.app.handlers.load_post.append(active_object_load_handler)


def set_object_visibility(object: bpy.types.Object, visible: bool) -> None:
    """Set visibility of Blender object"""
    if object.name not in bpy.context.view_layer.objects:
        return
    try:
        object.hide_set(not visible)
        # obj.hide_viewport = hide
        # icon incorrect, but causes blender bug for volumes with above
        object.hide_render = not visible
    except:
        # Keyframing object visibility can at times lead to:
        # RuntimeError: Object can't be hidden because it is not in View Layer 'ViewLayer'!
        pass
