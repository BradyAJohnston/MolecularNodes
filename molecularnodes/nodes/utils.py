import bpy
import databpy


def get_links(socket: bpy.types.NodeSocket) -> bpy.types.NodeLinks:
    """
    Get the links between two sockets
    """
    links = socket.links
    if links is None:
        raise ValueError("No links found for socket")
    return links


def int_menu_switch_from_dict(
    dictionary: dict, name: str = "NewMenuSwitch"
) -> bpy.types.NodeTree:
    """
    Create a menu from a dictionary of options
    """
    tree = databpy.nodes.new_tree(name=name, geometry=False)

    node_m = tree.nodes.new("GeometryNodeMenuSwitch")
    node_m.data_type = "INT"

    starting_enum_length = len(node_m.enum_items)

    for i, (key, value) in enumerate(dictionary.items()):
        if i < starting_enum_length:
            node_m.enum_items[i].name = key
        else:
            node_m.enum_items.new(name=key)

        node_m.inputs[i + 1].default_value = value

    tree.interface.new_socket("Selector", socket_type="NodeSocketMenu")
    tree.interface.new_socket("Value", socket_type="NodeSocketInt", in_out="OUTPUT")
    tree.links.new(databpy.nodes.get_input(tree).outputs["Selector"], node_m.inputs[0])
    tree.links.new(
        node_m.outputs["Output"], databpy.nodes.get_output(tree).inputs["Value"]
    )
    return tree
