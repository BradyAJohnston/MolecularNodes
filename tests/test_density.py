import molecularnodes as mn
import bpy
import pytest
import numpy as np
import itertools
from .constants import data_dir
from .utils import NumpySnapshotExtension
from databpy import ObjectTracker

try:
    import pyopenvdb
except ImportError:
    pytest.skip("pyopenvdb not installed", allow_module_level=True)


@pytest.fixture
def density_file():
    file = data_dir / "emd_24805.map.gz"
    vdb_file = data_dir / "emd_24805.vdb"
    vdb_file.unlink(missing_ok=True)
    # Make all densities are removed
    for o in bpy.data.objects:
        if o.mn.entity_type == "density":
            bpy.data.objects.remove(o, do_unlink=True)
    return file


def test_density_load(density_file):
    density = mn.entities.density.load(density_file)
    pos = density.named_attribute("position")

    assert len(pos) > 1000

    avg = np.mean(pos, axis=0)
    assert np.linalg.norm(avg) > 0.5

    print(f"{list(bpy.data.objects)=}")

    assert density.object.mn.entity_type == "density"
    assert density.object.users_collection[0] == mn.blender.coll.mn()


def test_density_centered(density_file):
    # First load using standard parameters to test recreation of vdb
    density = mn.entities.density.load(density_file, center=True, overwrite=True)
    pos = density.named_attribute("position")
    avg = np.mean(pos, axis=0)
    assert len(pos) > 1000
    assert np.linalg.norm(avg) < 0.1


def test_density_invert(density_file):
    # First load using standard parameters to test recreation of vdb
    density = mn.entities.density.load(density_file)

    # Then refresh the scene
    bpy.data.objects.remove(density.object, do_unlink=True)

    density = mn.entities.density.load(density_file, invert=True)
    style_node = mn.blender.nodes.get_style_node(density.object)
    style_node.inputs["Threshold"].default_value = 0.01

    pos = density.named_attribute("position")
    # At this threshold after inverting we should have a cube the size of the volume
    assert pos[:, 0].max() > 2.0
    assert pos[:, 1].max() > 2.0
    assert pos[:, 2].max() > 2.0


def test_density_multiple_load():
    file = data_dir / "emd_24805.map.gz"
    density1 = mn.entities.density.load(file)
    density2 = mn.entities.density.load(file)

    assert density1.object.mn.entity_type == "density"
    assert density2.object.mn.entity_type == "density"
    assert density1.object.users_collection[0] == mn.blender.coll.mn()
    assert density2.object.users_collection[0] == mn.blender.coll.mn()


@pytest.mark.parametrize("name", ["", "NewDensity"])
def test_density_naming_op(density_file, name):
    bpy.context.scene.MN_import_density_name = name
    bpy.context.scene.MN_import_density = str(density_file)
    bpy.ops.mn.import_density()

    if name == "":
        object_name = "emd_24805"
    else:
        object_name = name
    object = bpy.data.objects[object_name]
    assert object
    assert object.name == object_name


@pytest.mark.parametrize("name", ["", "NewDensity"])
def test_density_naming_api(density_file, name):
    density = mn.entities.density.load(density_file, name)

    if name == "":
        object_name = "emd_24805"
    else:
        object_name = name

    assert density.object
    assert density.object.name == object_name


@pytest.mark.parametrize(
    "invert,node_setup,center", list(itertools.product([True, False], repeat=3))
)
def test_density_operator(
    snapshot_custom: NumpySnapshotExtension, density_file, invert, node_setup, center
):
    scene = bpy.context.scene
    scene.MN_import_density = str(density_file)
    scene.MN_import_density_invert = invert
    scene.MN_import_node_setup = node_setup
    scene.MN_import_density_center = center
    scene.MN_import_density_name = ""
    with ObjectTracker() as o:
        bpy.ops.mn.import_density()
        density = scene.MNSession.match(o.latest())

    assert snapshot_custom == density.position
