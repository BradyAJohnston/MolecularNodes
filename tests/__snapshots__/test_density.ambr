# serializer version: 1
# name: test_density_operator[False-False-False]
  []
# ---
# name: test_density_operator[False-False-True]
  []
# ---
# name: test_density_operator[False-True-False]
  [[1.2 1.5 1.2]
   [0.9 1.0 1.5]
   [1.2 1.5 1.0]
   [1.3 0.7 1.0]
   [0.8 0.9 1.3]
   [0.9 0.8 1.0]
   [1.0 0.9 1.1]
   [1.3 1.3 1.4]
   [1.2 1.9 1.2]
   [1.2 1.6 0.9]
   [1.3 1.8 1.3]
   [1.2 0.9 1.4]
   [1.1 0.9 1.2]
   [1.1 0.7 1.0]
   [1.3 0.9 1.0]
   [1.2 1.6 1.0]
   [1.1 1.6 1.1]
   [1.4 1.5 1.4]
   [1.2 1.0 1.4]
   [1.1 1.4 1.1]
   [1.1 1.4 1.2]
   [1.3 1.6 1.3]
   [1.0 0.7 1.1]
   [1.0 1.2 1.3]
   [1.2 1.3 1.2]
   [1.2 1.2 1.1]
   [1.3 2.1 1.3]
   [1.4 1.6 1.3]
   [1.3 0.9 0.8]
   [1.3 1.4 1.1]
   [1.2 1.4 1.1]
   [1.3 1.7 1.1]
   [0.9 0.6 1.1]
   [1.3 0.9 1.3]
   [1.1 1.0 1.4]
   [1.0 0.7 1.3]
   [1.2 0.9 1.1]
   [1.2 1.0 1.1]
   [1.3 0.9 0.8]
   [1.4 1.0 1.1]
   [1.3 2.1 1.2]
   [0.9 0.8 1.1]
   [1.5 1.5 1.1]
   [1.1 1.5 1.2]
   [1.4 1.1 1.1]
   [1.3 1.8 1.2]
   [1.3 0.9 1.0]
   [1.0 0.8 1.3]
   [0.8 0.8 1.3]
   [1.2 2.0 1.2]
   [1.1 0.6 0.9]
   [1.3 1.3 1.3]
   [1.2 0.9 1.5]
   [1.2 1.6 0.9]
   [1.0 1.2 1.5]
   [1.4 1.5 1.1]
   [1.2 1.5 1.2]
   [1.2 2.1 1.2]
   [1.3 2.0 1.1]
   [1.3 1.6 1.0]
   [1.0 0.8 1.0]
   [1.2 1.8 1.2]
   [1.2 0.9 1.1]
   [1.3 0.7 0.9]
   [1.2 1.3 1.1]
   [1.3 1.0 1.2]
   [1.1 0.6 1.1]
   [1.2 0.8 1.3]
   [1.2 2.0 1.2]
   [1.0 1.0 1.3]
   [1.0 0.6 1.2]
   [1.3 1.4 1.1]
   [1.3 1.6 1.4]
   [1.1 1.2 1.2]
   [1.0 1.6 1.2]
   [1.0 1.0 1.4]
   [1.2 1.3 1.2]
   [1.3 1.3 1.4]
   [1.3 0.9 0.9]
   [1.1 0.8 1.2]
   [1.4 1.4 1.3]
   [1.1 1.4 1.2]
   [1.1 1.0 1.5]
   [1.2 1.5 1.1]
   [1.1 0.7 0.9]
   [1.1 0.8 1.4]
   [1.2 1.3 1.1]
   [1.1 0.8 0.8]
   [1.3 2.0 1.2]
   [1.1 1.1 1.4]
   [0.9 0.9 0.9]
   [1.1 1.1 1.1]
   [0.9 1.2 1.2]
   [1.0 0.6 0.9]
   [1.3 1.1 1.3]
   [1.1 1.0 1.3]
   [1.1 1.3 1.2]
   [1.3 1.5 1.1]
   [1.2 1.5 1.3]
   [1.2 1.6 1.2]]
# ---
# name: test_density_operator[False-True-True]
  [[ 0.0e+00  3.3e-01  3.5e-02]
   [-3.1e-01 -2.3e-01  2.7e-01]
   [ 2.5e-02  3.2e-01 -2.4e-01]
   [ 1.2e-01 -4.9e-01 -2.0e-01]
   [-4.0e-01 -3.3e-01  1.0e-01]
   [-2.9e-01 -3.9e-01 -2.4e-01]
   [-1.7e-01 -3.1e-01 -5.4e-02]
   [ 1.3e-01  8.3e-02  1.9e-01]
   [-1.8e-03  6.6e-01  3.2e-02]
   [-3.4e-02  3.8e-01 -2.5e-01]
   [ 1.5e-01  6.5e-01  1.1e-01]
   [-3.2e-02 -2.6e-01  1.9e-01]
   [-7.4e-02 -3.4e-01  1.7e-02]
   [-1.4e-01 -5.0e-01 -1.6e-01]
   [ 8.0e-02 -3.2e-01 -2.1e-01]
   [ 1.6e-02  4.2e-01 -1.7e-01]
   [-1.4e-01  3.9e-01 -8.0e-02]
   [ 2.0e-01  3.5e-01  2.3e-01]
   [ 3.9e-02 -1.8e-01  2.2e-01]
   [-1.0e-01  2.5e-01 -1.0e-01]
   [-8.9e-02  2.5e-01 -1.9e-02]
   [ 9.2e-02  4.2e-01  1.1e-01]
   [-2.3e-01 -4.5e-01 -9.5e-02]
   [-1.6e-01  2.1e-02  8.0e-02]
   [-6.4e-03  1.1e-01  2.0e-02]
   [-7.1e-03 -1.6e-04 -9.1e-02]
   [ 1.3e-01  8.6e-01  8.7e-02]
   [ 1.9e-01  3.6e-01  1.5e-01]
   [ 7.5e-02 -3.3e-01 -3.5e-01]
   [ 7.5e-02  1.6e-01 -1.1e-01]
   [ 3.0e-02  2.1e-01 -6.9e-02]
   [ 1.1e-01  5.5e-01 -5.0e-02]
   [-2.6e-01 -5.6e-01 -5.0e-02]
   [ 7.9e-02 -3.1e-01  8.8e-02]
   [-8.3e-02 -2.1e-01  2.2e-01]
   [-2.1e-01 -4.7e-01  9.8e-02]
   [-1.7e-02 -2.8e-01 -6.0e-02]
   [ 4.1e-02 -1.8e-01 -9.9e-02]
   [ 6.9e-02 -2.8e-01 -4.0e-01]
   [ 1.6e-01 -1.8e-01 -1.2e-01]
   [ 7.5e-02  8.9e-01  2.6e-03]
   [-2.5e-01 -4.3e-01 -1.1e-01]
   [ 3.1e-01  2.8e-01 -4.6e-02]
   [-1.3e-01  3.0e-01 -4.5e-02]
   [ 1.9e-01 -1.3e-01 -1.0e-01]
   [ 5.9e-02  6.5e-01  4.9e-02]
   [ 1.4e-01 -2.7e-01 -1.8e-01]
   [-2.3e-01 -4.3e-01  7.5e-02]
   [-3.6e-01 -3.6e-01  6.4e-02]
   [-1.0e-02  7.7e-01 -1.8e-02]
   [-7.9e-02 -6.0e-01 -3.3e-01]
   [ 8.3e-02  6.6e-02  6.6e-02]
   [-2.5e-02 -3.4e-01  2.9e-01]
   [ 2.5e-02  3.7e-01 -2.8e-01]
   [-1.9e-01  2.1e-02  2.6e-01]
   [ 1.7e-01  3.1e-01 -1.3e-01]
   [-1.7e-02  3.5e-01 -1.9e-04]
   [ 4.6e-02  8.9e-01 -2.9e-02]
   [ 1.1e-01  8.1e-01 -9.6e-02]
   [ 7.7e-02  3.7e-01 -2.2e-01]
   [-1.8e-01 -3.8e-01 -2.2e-01]
   [ 2.8e-02  6.1e-01 -2.4e-03]
   [ 0.0e+00 -3.0e-01 -8.2e-02]
   [ 5.8e-02 -4.5e-01 -2.5e-01]
   [ 8.3e-03  1.0e-01 -1.2e-01]
   [ 1.2e-01 -2.0e-01 -3.4e-02]
   [-1.2e-01 -5.9e-01 -5.6e-02]
   [-1.7e-02 -3.6e-01  9.1e-02]
   [-1.7e-02  8.1e-01 -1.1e-02]
   [-2.0e-01 -1.6e-01  1.0e-01]
   [-2.3e-01 -5.6e-01  3.3e-02]
   [ 8.7e-02  2.4e-01 -1.3e-01]
   [ 7.4e-02  3.7e-01  1.9e-01]
   [-1.2e-01 -3.3e-02 -2.4e-02]
   [-1.6e-01  3.6e-01  5.3e-02]
   [-2.3e-01 -2.4e-01  2.5e-01]
   [ 2.0e-02  1.4e-01  4.2e-02]
   [ 1.3e-01  8.3e-02  1.9e-01]
   [ 9.9e-02 -3.4e-01 -2.6e-01]
   [-1.2e-01 -4.4e-01  4.8e-02]
   [ 1.8e-01  2.1e-01  8.7e-02]
   [-1.4e-01  2.4e-01 -3.2e-02]
   [-6.2e-02 -1.5e-01  2.6e-01]
   [ 1.7e-02  2.5e-01 -1.3e-01]
   [-6.0e-02 -5.1e-01 -3.3e-01]
   [-1.4e-01 -4.3e-01  2.3e-01]
   [ 1.1e-02  5.8e-02 -8.1e-02]
   [-9.1e-02 -4.2e-01 -3.5e-01]
   [ 8.3e-02  8.3e-01 -1.1e-02]
   [-1.2e-01 -7.9e-02  1.6e-01]
   [-2.5e-01 -2.8e-01 -2.7e-01]
   [-1.4e-01 -5.9e-02 -5.8e-02]
   [-2.6e-01 -1.8e-02 -4.0e-02]
   [-1.5e-01 -5.5e-01 -2.8e-01]
   [ 1.4e-01 -1.2e-01  1.3e-01]
   [-9.7e-02 -2.1e-01  9.1e-02]
   [-5.0e-02  5.7e-02 -3.6e-02]
   [ 7.9e-02  2.8e-01 -4.7e-02]
   [ 2.9e-02  3.5e-01  1.0e-01]
   [ 4.1e-02  3.8e-01 -3.7e-02]]
# ---
# name: test_density_operator[True-False-False]
  []
# ---
# name: test_density_operator[True-False-True]
  []
# ---
# name: test_density_operator[True-True-False]
  [[0.8 0.9 1.5]
   [1.2 1.3 1.2]
   [1.3 0.8 0.8]
   [1.0 0.9 1.0]
   [0.8 0.7 0.9]
   [1.0 0.7 1.1]
   [1.2 1.5 1.2]
   [1.2 0.9 1.1]
   [1.1 0.9 1.6]
   [1.0 0.9 0.9]
   [1.3 0.9 0.8]
   [1.3 1.4 1.2]
   [1.0 1.0 1.6]
   [1.2 1.2 1.0]
   [1.4 1.4 1.2]
   [1.4 1.5 1.1]
   [0.8 0.7 1.3]
   [1.0 0.9 1.6]
   [1.2 0.9 1.1]
   [0.8 0.6 1.2]
   [1.3 1.4 1.0]
   [1.2 0.7 1.5]
   [0.8 0.7 1.1]
   [1.1 1.1 1.7]
   [1.2 1.4 1.1]
   [1.2 0.8 0.9]
   [1.1 1.4 1.3]
   [0.8 0.7 1.2]
   [1.0 1.0 1.4]
   [1.4 1.0 1.0]
   [0.8 0.8 1.3]
   [0.7 0.7 1.2]
   [1.0 1.0 1.3]
   [1.2 1.0 1.3]
   [1.3 1.3 1.1]
   [1.1 0.9 1.6]
   [1.3 1.5 1.3]
   [1.3 1.1 1.2]
   [1.1 0.7 1.1]
   [1.2 1.0 1.5]
   [1.1 1.0 1.1]
   [1.4 1.1 1.0]
   [1.1 0.8 1.2]
   [1.1 1.0 1.1]
   [0.9 0.7 1.2]
   [0.9 0.5 1.3]
   [1.2 1.8 1.4]
   [1.3 1.2 1.2]
   [1.0 1.1 1.1]
   [0.9 0.6 1.3]
   [1.2 1.0 1.0]
   [1.1 0.8 1.5]
   [1.1 1.0 1.3]
   [1.2 1.0 1.4]
   [1.2 1.2 1.1]
   [1.0 1.3 1.2]
   [0.9 1.0 1.4]
   [0.9 1.1 1.5]
   [1.3 1.9 1.2]
   [0.9 0.9 1.0]
   [1.0 0.9 1.1]
   [1.0 1.0 1.3]
   [1.0 0.7 1.0]
   [1.1 0.7 1.2]
   [1.1 0.8 1.3]
   [1.2 0.9 1.4]
   [1.2 1.0 1.1]
   [1.3 1.3 1.1]
   [1.3 0.9 1.3]
   [1.4 1.0 0.8]
   [0.9 0.8 1.2]
   [1.3 1.4 1.1]
   [0.7 0.8 1.4]
   [1.0 0.9 1.2]
   [1.3 1.2 1.0]
   [1.3 1.0 1.3]
   [0.8 0.8 1.1]
   [1.1 1.2 1.2]
   [1.2 0.6 0.8]
   [0.9 0.6 1.2]
   [1.1 1.4 1.2]
   [1.1 1.5 1.1]
   [1.2 1.0 1.4]
   [0.8 0.8 1.5]
   [1.3 1.5 1.2]
   [1.3 1.5 1.0]
   [1.2 1.6 1.2]
   [1.1 0.7 1.6]
   [1.2 1.3 1.0]
   [0.8 1.0 1.5]
   [1.1 0.9 1.4]
   [1.2 0.9 1.0]
   [1.1 1.8 1.3]
   [1.2 1.4 1.1]
   [1.0 0.5 1.3]
   [1.0 0.8 1.0]
   [1.0 0.9 0.9]
   [1.4 1.3 1.1]
   [0.9 0.8 1.3]
   [1.1 0.8 0.9]]
# ---
# name: test_density_operator[True-True-True]
  [[-0.4 -0.3  0.3]
   [ 0.0  0.1  0.1]
   [ 0.1 -0.4 -0.4]
   [-0.2 -0.3 -0.2]
   [-0.3 -0.5 -0.3]
   [-0.2 -0.5 -0.1]
   [-0.0  0.3  0.0]
   [ 0.0 -0.3 -0.0]
   [-0.1 -0.3  0.4]
   [-0.2 -0.3 -0.3]
   [ 0.1 -0.3 -0.4]
   [ 0.1  0.2 -0.0]
   [-0.2 -0.2  0.4]
   [ 0.0 -0.0 -0.2]
   [ 0.2  0.2 -0.0]
   [ 0.2  0.3 -0.1]
   [-0.4 -0.5  0.1]
   [-0.2 -0.3  0.4]
   [ 0.0 -0.3 -0.1]
   [-0.4 -0.6 -0.0]
   [ 0.1  0.2 -0.2]
   [-0.0 -0.5  0.3]
   [-0.4 -0.5 -0.1]
   [-0.1 -0.1  0.5]
   [-0.0  0.2 -0.1]
   [-0.0 -0.4 -0.3]
   [-0.1  0.2  0.1]
   [-0.4 -0.5 -0.0]
   [-0.1 -0.2  0.2]
   [ 0.2 -0.2 -0.2]
   [-0.4 -0.4  0.1]
   [-0.5 -0.5  0.0]
   [-0.2 -0.2  0.1]
   [-0.0 -0.2  0.2]
   [ 0.1  0.1 -0.1]
   [-0.1 -0.3  0.4]
   [ 0.1  0.3  0.1]
   [ 0.1 -0.1  0.0]
   [-0.1 -0.5 -0.1]
   [-0.0 -0.2  0.3]
   [-0.1 -0.2 -0.1]
   [ 0.2 -0.1 -0.2]
   [-0.1 -0.4 -0.0]
   [-0.1 -0.2 -0.1]
   [-0.3 -0.5  0.0]
   [-0.3 -0.7  0.1]
   [ 0.0  0.6  0.2]
   [ 0.1 -0.0 -0.0]
   [-0.2 -0.1 -0.1]
   [-0.3 -0.6  0.1]
   [ 0.0 -0.2 -0.2]
   [-0.1 -0.4  0.3]
   [-0.1 -0.2  0.1]
   [ 0.0 -0.2  0.2]
   [ 0.0  0.0 -0.1]
   [-0.1  0.1  0.0]
   [-0.3 -0.2  0.2]
   [-0.3 -0.1  0.3]
   [ 0.1  0.7 -0.0]
   [-0.3 -0.3 -0.2]
   [-0.2 -0.3 -0.1]
   [-0.2 -0.2  0.1]
   [-0.2 -0.5 -0.2]
   [-0.1 -0.5 -0.0]
   [-0.1 -0.4  0.2]
   [ 0.0 -0.3  0.2]
   [ 0.0 -0.2 -0.1]
   [ 0.1  0.1 -0.1]
   [ 0.1 -0.3  0.1]
   [ 0.2 -0.2 -0.4]
   [-0.3 -0.4  0.0]
   [ 0.1  0.2 -0.1]
   [-0.5 -0.4  0.2]
   [-0.2 -0.3  0.0]
   [ 0.1 -0.0 -0.2]
   [ 0.1 -0.2  0.1]
   [-0.4 -0.4 -0.1]
   [-0.1  0.0  0.0]
   [-0.0 -0.6 -0.4]
   [-0.3 -0.6  0.0]
   [-0.1  0.2 -0.0]
   [-0.1  0.3 -0.1]
   [ 0.0 -0.2  0.2]
   [-0.4 -0.4  0.3]
   [ 0.1  0.3  0.0]
   [ 0.1  0.3 -0.2]
   [-0.0  0.4 -0.0]
   [-0.1 -0.5  0.4]
   [-0.0  0.1 -0.2]
   [-0.4 -0.2  0.3]
   [-0.1 -0.3  0.2]
   [ 0.1 -0.3 -0.2]
   [-0.1  0.6  0.1]
   [ 0.0  0.2 -0.1]
   [-0.2 -0.7  0.1]
   [-0.2 -0.4 -0.2]
   [-0.2 -0.3 -0.3]
   [ 0.2  0.1 -0.1]
   [-0.3 -0.4  0.1]
   [-0.1 -0.4 -0.3]]
# ---
